fairlearn/postprocessing/_threshold_optimizer.py

```diff
@@ -427,6 +260,167 @@ def _threshold_optimization_equalized_odds(sensitive_features, labels, scores, g
+        threshold_optimizer = _SUPPORTED_CONSTRAINTS[self.constraints]
+
+        if not is_classifier(self.estimator):
+            raise ValueError("{} needs a classifier to work on, and {} is not "
+                             "one.".format(self.__class__.__name__,
+                                           self.estimator.__class__.__name__))
+
+        X, y, sensitive_features = self._validate_input_data(
+            X, sensitive_features, y)
+
+        self.estimator_ = self.estimator.fit(X, y, **kwargs)
+
+        scores = _get_soft_predictions(self.estimator_, X)
+        self._post_processed_predictor_by_sensitive_feature = threshold_optimizer(
+            sensitive_features, y, scores, self.grid_size, self.flip, self.plot)
+
+    def predict(self, X, *, sensitive_features):
+        """Predict label for each sample in X while taking into account sensitive features.
+
+        :param X: feature matrix
+        :type X: numpy.ndarray or pandas.DataFrame
+
+        :param sensitive_features: sensitive features to identify groups by, currently allows
+            only a single column
+        :type sensitive_features: currently 1D array as numpy.ndarray, list, pandas.DataFrame,
+            or pandas.Series
+
+        :return: predictions in numpy.ndarray
+        """
+        check_is_fitted(self)
+        random_state = check_random_state(self.random_state)
+
+        X, _, sensitive_features = self._validate_input_data(
+            X, sensitive_features)
+
+        unconstrained_predictions = self._unconstrained_predictor.predict(X)
+
+        positive_probs = _vectorized_prediction(
+            self._post_processed_predictor_by_sensitive_feature,
+            sensitive_features,
+            unconstrained_predictions)
+        return (positive_probs >= random_state.rand(len(positive_probs))) * 1
+
+    def _pmf_predict(self, X, *, sensitive_features):
+        """Probabilistic mass function.
+
+        :param X: Feature matrix
+        :type X: numpy.ndarray or pandas.DataFrame
+
+        :param sensitive_features: Sensitive features to identify groups by,
+            currently allows only a single column
+
+        :type sensitive_features: Currently 1D array as numpy.ndarray, list,
+            pandas.DataFrame, or pandas.Series
+
+        :return: array of tuples with probabilities for predicting 0 or 1,
+            respectively. The sum of the two numbers in each tuple needs to add
+            up to 1.
+        :rtype: numpy.ndarray
+        """
+        check_is_fitted(self)
+        X, _, sensitive_features = self._validate_input_data(
+            X, sensitive_features)
+        positive_probs = _vectorized_prediction(
+            self._post_processed_predictor_by_sensitive_feature, sensitive_features,
+            self._unconstrained_predictor.predict(X))
+        return np.array([[1.0 - p, p] for p in positive_probs])
+
+    def _validate_input_data(self, X, sensitive_features, y=None):
```

Comment:
We'll have some major merge conflicts here with my PR #280 . Just an FYI. I'll probably take until tomorrow to finish it even in the current state, but maybe it's worth incorporating your changes first and then I'll adjust mine since I've unified the input validation for reductions and postprocessing. I love that this allows us to not reinvent the wheel and just use sklearn's input validation which should be more obvious to users anyway. So in summary: I'd like your change to go in first, if possible, and I'll unify them then

Reply:
What we could do, is to put both this PR and #280 on hold, and have a separate PR which does the input validation change only. It'll also make #280 probably much smaller, since that's a pretty big one, and ideally could be broken into a few PRs, which makes the review also easier. WDYT?

Reply:
My change influences input validation pretty heavily since I'm allowing extra inputs, namely multi-dimensional sensitive features. I'd like to get mine in asap, so depending whether you think yours can go in I can resolve merge conflicts in my branch after you merge, or I can resolve them on yours after I merge.

Reply:
I'm happy with how everything looks here. Any chance this can be checked in this week? That way I could resolve the conflicts in my branch which will be major if we make more adjustments to the tests.

Reply:
I may go ahead and check mine in first, but if that happens I'll create a branch with test changes required for yours, so you can merge with it to avoid having to fix all the merge conflicts.

Reply:
Sure, go ahead. I may break this PR to a few different ones to make it easier to review anyway.

Reply:
Since this has happened now I can close this thread since it was just meant as a warning